src/WebJobs.Script/Host/ProxyFunctionProvider.cs

```diff
@@ -0,0 +1,195 @@
+
+            _fileChangeSubscription = eventManager.OfType<FileEvent>()
+                       .Where(f => string.Equals(f.Source, EventSources.ScriptFiles, StringComparison.Ordinal) &&
+                       string.Equals(Path.GetFileName(f.FileChangeArguments.Name), ScriptConstants.ProxyMetadataFileName, StringComparison.OrdinalIgnoreCase))
+                       .Subscribe(e => HandleProxyFileChange());
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
+
+        public Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync()
+        {
+            _metadataLock.EnterReadLock();
+            try
+            {
+                return Task.FromResult(_metadata.Value);
+            }
+            finally
+            {
+                _metadataLock.ExitReadLock();
+            }
+        }
+
+        private void HandleProxyFileChange()
+        {
+            _metadataLock.EnterWriteLock();
+            try
+            {
+                if (_metadata.IsValueCreated)
+                {
+                    _metadata = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
+                }
+            }
+            finally
+            {
+                _metadataLock.ExitWriteLock();
+            }
+        }
+
+        private ImmutableArray<FunctionMetadata> LoadFunctionMetadata()
+        {
+            var functionErrors = new Dictionary<string, ICollection<string>>();
+            Collection<ProxyFunctionMetadata> proxies = ReadProxyMetadata(_scriptOptions.Value.RootScriptPath, _logger, functionErrors);
+
+            ImmutableArray<FunctionMetadata> metadata;
+            if (proxies != null && proxies.Any())
+            {
+                metadata = proxies.ToImmutableArray<FunctionMetadata>();
+            }
+
+            _functionErrors = functionErrors;
+            return metadata;
+        }
+
+        internal static Collection<ProxyFunctionMetadata> ReadProxyMetadata(string scriptPath, ILogger logger, Dictionary<string, ICollection<string>> functionErrors = null)
+        {
+            functionErrors = functionErrors ?? new Dictionary<string, ICollection<string>>();
+
+            // read the proxy config
+            string proxyConfigPath = Path.Combine(scriptPath, ScriptConstants.ProxyMetadataFileName);
+            if (!File.Exists(proxyConfigPath))
+            {
+                return null;
+            }
+
+            string proxiesJson = File.ReadAllText(proxyConfigPath);
+            if (!string.IsNullOrWhiteSpace(proxiesJson))
+            {
+                logger.LogInformation("Loading proxies metadata");
+                var metadataCollection = LoadProxyMetadata(proxiesJson, functionErrors, logger);
+                logger.LogInformation($"{metadataCollection.Count} proxies loaded");
```

Comment:
Consider using structured logging.

```suggestion
                logger.LogInformation("{ProxyCount} proxies loaded", metadataCollection.Count);
```